dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1051 @@
+		Hash{}, nil)
+	cs.blockState.(*syncmocks.BlockState).On("HasHeader", mock.AnythingOfType("common.Hash")).Return(true, nil)
+
+	err = cs.setPeerHead(testPeer, hash, number-1)
+	require.NoError(t, err)
+	expected = &peerState{
+		who:    testPeer,
+		hash:   hash,
+		number: number - 1,
+	}
+	require.Equal(t, expected, cs.peerState[testPeer])
+	select {
+	case <-cs.workQueue:
+		t.Fatal("should not put fork we already have into work queue")
+	default:
+	}
+}
+
+func TestChainSync_sync_bootstrap_withWorkerError(t *testing.T) {
+	cs, _ := newTestChainSync(t)
+
+	go cs.sync()
+	defer cs.cancel()
+
+	testPeer := peer.ID("noot")
+	cs.peerState[testPeer] = &peerState{
+		number: 1000,
+	}
+
+	cs.workQueue <- cs.peerState[testPeer]
+
+	select {
+	case res := <-cs.resultQueue:
+		expected := &workerError{
+			err: errNilResponse, // since MockNetwork returns a nil response
+			who: testPeer,
+		}
+		require.Equal(t, expected, res.err)
+	case <-time.After(testTimeout):
+		t.Fatal("did not get worker response")
+	}
+
+	require.Equal(t, bootstrap, cs.state)
+}
+
+func TestChainSync_sync_tip(t *testing.T) {
+	cs, _ := newTestChainSync(t)
+	cs.blockState = new(syncmocks.BlockState)
+	header, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, 1000,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*syncmocks.BlockState).On("BestBlockHeader").Return(header, nil)
+	cs.blockState.(*syncmocks.BlockState).On("GetHighestFinalisedHeader").Return(header, nil)
+
+	go cs.sync()
+	defer cs.cancel()
+
+	testPeer := peer.ID("noot")
+	cs.peerState[testPeer] = &peerState{
+		number: 999,
+	}
+
+	cs.workQueue <- cs.peerState[testPeer]
+	time.Sleep(time.Second)
+	require.Equal(t, tip, cs.state)
+}
+
+func TestChainSync_sync_resultQueue(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	cs, _ := newTestChainSync(t)
+	cs.blockState = new(syncmocks.BlockState)
+	header, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, 1000,
+		types.NewDigest())
+	require.NoError(t, err)
+	cs.blockState.(*syncmocks.BlockState).On("BestBlockHeader").Return(header, nil)
+	cs.blockState.(*syncmocks.BlockState).On("GetHighestFinalisedHeader").Return(header, nil)
+	mockWorkHandler := NewMockworkHandler(ctrl)
+	cs.handler = mockWorkHandler
+	mockWorkHandler.EXPECT().handleWorkerResult(gomock.AssignableToTypeOf(&worker{})).Return(&worker{}, nil)
+	mockWorkHandler.EXPECT().hasCurrentWorker(gomock.AssignableToTypeOf(&worker{}),
+		gomock.AssignableToTypeOf(map[uint64]*worker{}))
+
+	go cs.sync()
+	defer cs.cancel()
+
+	w := worker{
+		ctx: context.Background(),
+		err: &workerError{
+			err: errors.New("protocol not supported"),
+		},
+	}
+
+	cs.resultQueue <- &w
+	time.Sleep(time.Second)
+
+	w = worker{
+		ctx: context.Background(),
+		err: &workerError{
+			err: context.DeadlineExceeded,
+		},
+	}
+
+	cs.resultQueue <- &w
+	time.Sleep(time.Second)
```

Comment:
I was looking into why you needed this sleep.  To make this code more testable, I'd suggest taking the entire case of `case res := <-cs.resultQueue:` in `chainSync.sync()` into another receiver function, then you can test that function on its own. You should be able to achieve more coverage as well.

Reply:
Good idea, updated.